bip-vaults.mediawiki

```diff
@@ -0,0 +1,757 @@
+<pre>
+  BIP: xxxx
+  Layer: Consensus (soft fork)
+  Title: OP_VAULT
+  Author: James O'Beirne <vaults@au92.org>
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-xxxx
+  Status: Draft
+  Type: Standards Track
+  Created: 2023-02-03
+  License: BSD-3-Clause
+  Post-History: 2023-01-09: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-January/021318.html [bitcoin-dev] OP_VAULT announcment
+</pre>
+
+
+== Introduction ==
+
+This BIP proposes new tapscript opcodes, <code>OP_VAULT</code> and
+<code>OP_UNVAULT</code>, that add consensus support for a specialized covenant. 
+These opcodes allow users to enforce a delay period before designated coins may
+be spent to an arbitrary destination, with the exception of a prespecified
+"recovery" path. At any time prior to final withdrawal, the coins can be spent to
+the prespecified path.
+
+
+=== Motivation ===
+
+The hazard of custodying Bitcoin is well known. Users of Bitcoin must go to
+significant effort to secure their private keys, and hope that once provisioned
+their custody system does not yield to any number of evolving and
+persistent threats. Users have little means to intervene once a compromise is
+detected. This proposal introduces a mechanism that significantly
+mitigates the worst-case outcome of key compromise: coin loss.
+
+Introducing a way to intervene during unexpected spends allows users to
+incorporate highly secure key storage methods or unusual fallback strategies
+that are only exercised in the worst case, and which may otherwise be
+operationally prohibitive. The goal of this proposal is to make this strategy
+usable for custodians of any size with minimal complication.
+
+==== Example uses ====
+
+An individual custodying Bitcoin uses the common "single signature and
+passphrase" configuration with a hardware wallet. They are concerned about the
+risk associated with relying on a single manufacturer for key management as
+well as physical access to the hardware, so they generate a new key that is
+highly secure, but would be impractical for daily use. For example the key
+could be generated in some analog fashion, or on an old computer (with added
+entropy) that is then destroyed, with the private key replicated only in paper
+form. Or the key could be a 2-of-3 multisig using devices from different
+manufacturers. Perhaps the key is geographically distributed.
+
+This individual can use <code>OP_VAULT</code> to make use of the highly secure
+key as the unlikely recovery path, while using their existing signing procedure
+as the withdrawal trigger key, with a configured spend delay of 1 day. They can
+run software on their mobile device that monitors the blockchain for spends of
+the vault outpoints.
+
+If the vaulted coins move in an unexpected way, the user can immediately sweep
+them to the highly secure recovery path, but spending the coins on a daily
+basis works in the same way it did prior to vaulting - aside from the spend
+delay.
```

Comment:
Completely handwaving here, but I think you'd need something like recursive taproot(or better bitcoin script) to do this properly. The "forwarded" script would at spend time get a `<pubkey>` that is inserted to a timelocked sub-expression in a larger expression that has otherwise been committed to at deposit time.

You can't just insert random bitcoin script at spend time, as we need to ensure it's purely restricting spends, i.e. cannot be OP_SUCCESSX or something that modifies the stack in an unexpected way.

Recipients would have to scan on the input side to decipher outputs have hit their wallet, and they would still have to sweep that output to call it a proper deposit, ala HTLC resolution.

I'm not sure that's a compelling use-case, since wallets will likely sweep outputs promptly for this very reason regardless.